src/toxcore/onion/onion_announce.rs

```diff
@@ -0,0 +1,521 @@
+/*
+    Copyright (C) 2013 Tox project All Rights Reserved.
+    Copyright © 2018 Evgeny Kurnevsky <kurnevsky@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+use std::cmp::Ordering;
+use std::net::IpAddr;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use toxcore::binary_io::*;
+use toxcore::crypto_core::*;
+use toxcore::onion::packet::*;
+use toxcore::dht::kbucket::Distance;
+
+/// Number of secret random bytes to make onion ping id unique for each node.
+pub const SECRET_BYTES_SIZE: usize = 32;
+
+/// Maximum number of entries in onion announce list. When number of entries
+/// exceeds this value farthest nodes are dropped using DHT distance function.
+pub const ONION_ANNOUNCE_MAX_ENTRIES: usize = 160;
+
+/// Number of seconds when onion ping id is valid after it was generated.
+/// To be precise ping id will be valid between `PING_ID_TIMEOUT` and
```

Comment:
What about `now() ..= 2* PING_ID_TIMEOUT`?

Reply:
Didn't get your question. What about it?

Reply:
Is it valid since `now()` till `2* PING_ID_TIMEOUT` or since `PING_ID_TIMEOUT` till `2* PING_ID_TIMEOUT`?

Reply:
Since now, reworded it.